flow/controllers/base_controller.py

```diff
@@ -235,4 +285,93 @@ def safe_velocity(self, env):
+                "=====================================".format(self.veh_id))
+
         return v_safe
+
+    def get_obey_speed_limit_action(self, env, action):
+        """Perform the "obey_speed_limit" failsafe action.
+
+        Checks if the computed acceleration would put us above edge speed limit.
+        If it would, output the acceleration that would put at the speed limit
+        velocity.
+
+        Parameters
+        ----------
+        env : flow.envs.Env
+            current environment, which contains information of the state of the
+            network at the current time step
+        action : float
+            requested acceleration action
+
+        Returns
+        -------
+        float
+            the requested action clipped by the speed limit
+        """
+        # check for speed limit
+        this_edge = env.k.vehicle.get_edge(self.veh_id)
+        edge_speed_limit = env.k.network.speed_limit(this_edge)
+
+        this_vel = env.k.vehicle.get_speed(self.veh_id)
+        sim_step = env.sim_step
+
+        if this_vel + action * sim_step > edge_speed_limit:
+            if edge_speed_limit > 0:
+                print(
+                    "=====================================\n"
+                    "Speed of vehicle {} is greater than speed limit. Obey "
+                    "speed limit clipping applied.\n"
+                    "=====================================".format(self.veh_id))
+                return (edge_speed_limit - this_vel) / sim_step
+            else:
+                return -this_vel / sim_step
+        else:
+            return action
+
+    def get_feasible_action(self, action):
+        """Perform the "feasible_accel" failsafe action.
+
+        Checks if the computed acceleration would put us above maximum
+        acceleration or deceleration. If it would, output the acceleration
+        equal to maximum acceleration or deceleration.
+
+        Parameters
+        ----------
+        action : float
+            requested acceleration action
+
+        Returns
+        -------
+        float
+            the requested action clipped by the feasible acceleration or
+            deceleration.
+        """
+        if action > self.max_accel:
+            action = self.max_accel
+
+            print(
+                "=====================================\n"
+                "Acceleration of vehicle {} is greater than the max "
+                "acceleration. Feasible acceleration clipping applied.\n"
+                "=====================================".format(self.veh_id))
+
+        if action < -self.max_deaccel:
+            action = -self.max_deaccel
+
+            print(
+                "=====================================\n"
+                "Deceleration of vehicle {} is greater than the max "
+                "deceleration. Feasible acceleration clipping applied.\n"
+                "=====================================".format(self.veh_id))
```

Comment:
I don't know how often all these failsafe warnings are printed, but if it's every iteration maybe it would be useful to add a printing condition (`self.display_warnings` boolean...)?